### PR TITLE
Rewrite wasm-bindgen with updated interface types proposal

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is a pretty large scale rewrite of the internals of wasm-bindgen. No user-facing changes are expected as a result of this PR, but due to the scale of changes here it's likely inevitable that at least something will break. I'm hoping to get more testing in though before landing!

The purpose of this PR is to update wasm-bindgen to the current state of the interface types proposal. The wasm-bindgen tool was last updated when it was still called "WebIDL bindings" so it's been awhile! All support is now based on https://github.com/bytecodealliance/wasm-interface-types which defines parsers/binary format/writers/etc for wasm-interface types.

This is a pretty massive PR and unfortunately can't really be split up any more afaik. I don't really expect realistic review of all the code here (or commits), but some high-level changes are:

* Interface types now consists of a set of "adapter functions". The IR in wasm-bindgen is modeled the same way not.
* Each adapter function has a list of instructions, and these instructions work at a higher level than wasm itself, for example with strings.
* The wasm-bindgen tool has a suite of instructions which are specific to it and not present in the standard. (like before with webidl bindings)
* The anyref/multi-value transformations are now greatly simplified. They're simply "optimization passes" over adapter functions, removing instructions that are otherwise present. This way we don't have to juggle so much all over the place, and instructions always have the same meaning.

TODO:

* [x] re-add optimization to skip shim functions in JS
* [x] test raytrace example still works

Follow-up work

* [ ] Add tests about the output of the wasm-interface types section from Rust
* [ ] Add tests which consume wasm interface types and produce a JS polyfill

### CODE REVIEW

**Comment by alexcrichton** [comment]
Note that this change isn't quite ready to land yet, but I wanted to start getting some CI feedback to see how this fares. Additionally I would like to greatly expand the test coverage of the anyref pass, multivalue pass, and support for interface types. I plan on doing all that before landing this.

File: crates/multi-value-xform/tests/align.wat [hunk 1]
```diff
@@ -0,0 +1,37 @@
+;; @xform export "foo" (f64 i32 i64)
+
+(module
+  (global (mut i32) (i32.const 0))
+  (memory 1)
+
+  (func $foo (export "foo") (param i32))
+)
+
+(; CHECK-ALL:
```
Comment (fitzgen) [inline crates/multi-value-xform/tests/align.wat]: Thanks for adding this :+1:

File: crates/cli-support/src/js/binding.rs [hunk 2]
```diff
@@ -536,4 +315,845 @@ impl JsBuilder {
...
+                    }
+                }
+            }
+        }
+
+        Instruction::Standard(wit_walrus::Instruction::IntToWasm { trap: false, .. }) => {
+            js.typescript_required("number");
+            let val = js.pop();
+            js.assert_number(&val);
+            js.push(val);
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/cli-support/src/js/binding.rs]: I thought it worked out pretty well too!

I've left some comments, good idea as always heh

File: crates/cli-support/src/multivalue.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        to_xform.push((*id, 0, types));
+        slots.push(id);
+        return;
+    }
+
+    // If the last instruction is a `StoreRetptr`, then this must be an adapter
+    // which calls an imported function.
+    //
+    // FIXME(#1872) handle this
+    // if let Some(Instruction::StoreRetptr { .. }) = instructions.last() {}
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/cli-support/src/multivalue.rs]: Nah this was never exercised or supported before, so I don't think we need to handle this. Once we start writing more wasm-interface-types tests, though, I think we'll want to fill this out pretty shortly afterwards.

**Comment by alexcrichton** [comment]
Ok thanks for the review @fitzgen!

I think I've taken care of everything now, so I'm going to merge this and incrementally improve support for wasm interface types in subsequent PRs, hopefully as a bit more digestable.

File: crates/anyref-xform/tests/anyref-param-owned.wat [hunk 4]
```diff
@@ -0,0 +1,31 @@
+;; @xform export "foo" (anyref_owned)
+
+(module
+  (func $foo (export "foo") (param i32))
+  (func $alloc (export "__wbindgen_anyref_table_alloc") (result i32)
+    i32.const 0)
+  (func $dealloc (export "__wbindgen_anyref_table_dealloc") (param i32))
+)
+
+(; CHECK-ALL:
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/anyref-xform/tests/anyref-param-owned.wat]: I've been tweaking this as I've gone along for various crates, but it's starting to reach a steady state, yeah, so we should probably look into extracting it soon!

**Pull request merged**